### PR TITLE
Download chart as PNG image option

### PR DESCRIPTION
**Pull request by mmcfarland**
## Overview

All of the Nivo chart types implemented in the app render to Canvas. As
a result, we can generate a dataUrl for each canvas element and offer it
as a direct image download. A new button is available on every individual
chart (or chart group in the case of Waffle) which gives a menu to download
 the image as PNG.  

A few workarounds are needed to accomplish
this:

- A ref to a particular container element is required in order to
generically search for a child canvas element
- Waffle charts are really a sequence of three separate charts that
should be treated as one
- Nivo charts have a transparent background, and a solid color needs to
be applied before writing

As a result of these extra requirements, there is some manual
manipulation of the canvas element to represent enough chart context to
be meaningful. Some content, like a title or geography, is not included
on the canvas based on the chart type. In these cases, I've made the
filename the place where _all_ data shows up. As an enhancement, we may
need to include additional text directly on the chart canvas while
generating the image.

Connects #48 

### Notes

While all relevant information is included in the file name, it would be nice to ensure it is included in the image content directly. This is related to #27 and may not be possible in the current scope.

The CSV menu item was added, but it is not implemented.

More charts _could_ be combined (like Stacked), but it's not entirely clear if it's desirable. I'm interested in client feedback about the utility of downloading select charts instead of big chunks of them. They may be embedded in other media better by remaining per-geography or group.

## Testing Instructions

 * For multiple geographies, select multiple questions of each chart type
 * Check each chart or chart group has a new hover-only button for a download menu
 * Clicking the download PNG option saves a PNG to the local machine which includes a representative rendering of the selected chart. 
 * The generated file name includes enough context to be able to associate the chart with a question.

### CODE REVIEW

File: src/app/src/hooks/useRefs.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import React, { useRef } from 'react';
+
+// A custom hook that creates an array of refs equaling the number of array
+// elements passed to it. Useful for creating refs to be used inside of a loop
+// where useRef is not valid.
+const useRefs = length => {
```
Comment (TaiWilkin) [inline src/app/src/hooks/useRefs.js:6]: Very cool hook. Nice way of managing the useRefs for this.

File: src/app/src/components/DownloadMenu.jsx [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+            'canvas'
+        );
+
+        const filename = createFilename(question);
+
+        // Depending on the chart type, there may be many canvas elements.
+        // Coerce them into a single canvas, or if just one, use it directly.
+        const chartCanvas =
+            chartCanvases.length > 1
+                ? combineCanvases(chartCanvases, combineDir)
```
Comment (TaiWilkin) [inline src/app/src/components/DownloadMenu.jsx:49]: This is a great way to do this. It's easy to read.

**Review by TaiWilkin — APPROVED** [review]
This works well. I tested it in Chrome, Safari, and Firefox and was able to successfully download all three chart types. Personally, I think the smaller chart selection downloads are better (the way you have it set up now), but obviously client feedback will be the deciding factor there. 

The stacked charts are very close together, which is making the download button show up next to the item above it (see photo). It's a bit confusing which goes with which. The easiest way to resolve that is probably just to increase the spacing between the stacked charts. 
Screen Shot 2021-01-04 at 4 14 52 PM

**Comment by mmcfarland** [comment]
Good catch. I updated the padding on stacked charts to better separate the menu

image

**Pull request merged**